### PR TITLE
reject seed candidates not part of superclusters

### PR DESCRIPTION
**Pull request by kpedro88**
During relval production for CMSSW_9_3_0_pre5, a crash was observed in `PFECALSuperClusterAlgo`. An [example config+log](http://cms-unified.web.cern.ch/cms-unified/joblogs/sandhya_RVCMSSW_9_3_0_pre5TTbar_14TeV_Timing__2023D17noPUTiming_170906_152309_28/139/RecoFullGlobal_Timing_2023D17/1a3ab292-9416-11e7-a633-02163e00d7b3-0-3-logArchive/cmsRun1/) was provided, which let me get a better stacktrace:
```
#0  0x00007fffad5b4675 in reco::CaloCluster::setCorrectedEnergy (this=0x7fff940785f0, cenergy=0)
    at /cvmfs/cms.cern.ch/slc6_amd64_gcc630/cms/cmssw/CMSSW_9_3_0_pre5/src/DataFormats/CaloRecHit/interface/CaloCluster.h:112
#1  PFECALSuperClusterAlgo::buildSuperCluster (this=this@entry=0x7fffa1082568, seed=..., clusters=std::vector of length 153, capacity 256 = {...})
    at /uscms_data/d3/pedrok/phase2/hgcal/CMSSW_9_3_0_pre5/src/RecoEcal/EgammaClusterAlgos/src/PFECALSuperClusterAlgo.cc:494
#2  0x00007fffad5b6628 in PFECALSuperClusterAlgo::buildAllSuperClusters (this=this@entry=0x7fffa1082568,
    clusters=std::vector of length 153, capacity 256 = {...}, seedthresh=<optimized out>)
    at /uscms_data/d3/pedrok/phase2/hgcal/CMSSW_9_3_0_pre5/src/RecoEcal/EgammaClusterAlgos/src/PFECALSuperClusterAlgo.cc:328
#3  0x00007fffad5b704a in PFECALSuperClusterAlgo::run (this=this@entry=0x7fffa1082568)
    at /uscms_data/d3/pedrok/phase2/hgcal/CMSSW_9_3_0_pre5/src/RecoEcal/EgammaClusterAlgos/src/PFECALSuperClusterAlgo.cc:314
#4  0x00007fffa679b8f8 in PFECALSuperClusterProducer::produce (this=0x7fffa1082400, iEvent=..., iSetup=...)
    at /uscms_data/d3/pedrok/phase2/hgcal/CMSSW_9_3_0_pre5/src/RecoEcal/EgammaClusterProducers/src/PFECALSuperClusterProducer.cc:187
```

Using gdb, I found an apparent problem in frame 1:
```
(gdb) frame 1
#1  PFECALSuperClusterAlgo::buildSuperCluster (this=this@entry=0x7fffa1082568, seed=..., clusters=std::vector of length 153, capacity 256 = {...})
    at /uscms_data/d3/pedrok/phase2/hgcal/CMSSW_9_3_0_pre5/src/RecoEcal/EgammaClusterAlgos/src/PFECALSuperClusterAlgo.cc:494
494       new_sc.setCorrectedEnergy(corrSCEnergy);
(gdb) print new_sc
$1 = {<reco::CaloCluster> = {_vptr.CaloCluster = 0x7ffff3f37580 <vtable for reco::SuperCluster+16>, energy_ = 0, correctedEnergy_ = -1,
    correctedEnergyUncertainty_ = -1, position_ = {fCoordinates = {fX = -nan(0x8000000000000), fY = -nan(0x8000000000000), fZ = -nan(0x8000000000000)}},
    caloID_ = {_vptr.CaloID = 0x7ffff3f374a8 <vtable for reco::CaloID+16>, detectors_ = 0}, hitsAndFractions_ = std::vector of length 0, capacity 0,
    algoID_ = reco::CaloCluster::undefined, seedId_ = {static kDetOffset = 28, static kSubdetOffset = 25, id_ = 0}, flags_ = 0,
    static flagsMask_ = 268435455, static flagsOffset_ = 28}, seed_ = {core_ = {cachePtr_ = {_M_b = {_M_p = 0x0}}, processIndex_ = 0, productIndex_ = 0},
    key_ = 18446744073709551615}, clusters_ = {<edm::PtrVectorBase> = {<No data fields>}, <No data fields>},
  preshowerClusters_ = {<edm::PtrVectorBase> = {<No data fields>}, <No data fields>}, preshowerEnergy_ = 0, rawEnergy_ = 0, phiWidth_ = 0, etaWidth_ = 0,
  preshowerEnergy1_ = 0, preshowerEnergy2_ = 0}
(gdb) print clustered
$6 = std::vector of length 0, capacity 0
```

This leads to divide by 0 problems, so the cluster has nan coordinates and no energy. I also turned on the verbose output and saw this:
```
Dumping cluster detail  Passed seed: e = 5.18964 eta = -2.24025 phi = 2.99611
        Clustered cluster: e = 5.18964 eta = -2.24025 phi = 2.99611
    Non-Clustered cluster: e = 0.0662978 eta = -1.71095 phi = -1.38067
...
Dumping cluster detail  Passed seed: e = 0.0662978 eta = -1.71095 phi = -1.38067
    Non-Clustered cluster: e = 0.0662978 eta = -1.71095 phi = -1.38067
```

Somehow the algorithm ends up in a state where everything is a "non-clustered cluster", so the assumption that there's a "clustered cluster" to start the new supercluster is violated. Further description from @rovere:
>Ok, I have some more details about the crash.
We have an HGCAL cluster whose energy/uncalib_energy are: E: 3.21491
UE: 0.0662978, respectively.
Now, the logic of SC is that all clusters are first ordered by Et (in
this case Et uses the Energy to order the clusters), and this little
cluster sits quite high in the final vector since it has a ptfast of
1.1251, thanks to its relatively high energy.
After that, the clusters are stable_partitioned so that all clusters
above a threshold are promoted to be seeds for a possible SC. The
threshold is 1 GeV and is applied on Et (of ptfast, if you prefer) so
this little guy passes the selection and happily becomes a seed
candidate.
After that, the clusters are again stable_partitioned so that all
clusters within the dyn window of the mustache algorithm for the
specific seed under consideration appear first. The problem is that in
doing this partitioning, this little guy is rejected, i.e. it is not
part of the SC it is seeding! And that's because in the mustache
algorithm there is this nice line[1] that is using the
**uncalibrated** energy as input. In this, maybe extremely unlucky
case, its value is 0.066XXX which is below the allowed (and **not
protected**) value of 0.07943XX that makes the sqrt negative, hence
quite a bit of nan, and the following logic simply does not hold any
longer, since the seed itself is not part of the SC.
[1]
https://github.com/cms-sw/cmssw/blob/master/RecoEcal/EgammaCoreTools/src/Mustache.cc#L38

For now, we just discard such clusters to let the algorithm progress. Eventually, for HGCal we should set uncalibrated energy equal to calibrated energy, and consider the consistency of the clustering algorithms in terms of which quantity they use.

This PR is critical for HGCal TDR production, so will be backported to 93X.

attn: @cseez, @bendavid, @felicepantaleo

### CODE REVIEW

**Comment by kpedro88** [comment]
assign upgrade

**Comment by rovere** [comment]
@kpedro88 can we put it on hold for a little while? I'm still investigating if there are better solutions...

**Comment by kpedro88** [comment]
@rovere okay, but I will leave the PRs open so we can go as soon as possible if we decide on this solution

**Comment by kpedro88** [comment]
hold

**Comment by kpedro88** [comment]
please test

File: RecoEcal/EgammaClusterAlgos/src/PFECALSuperClusterAlgo.cc [hunk 1]
```diff
@@ -392,6 +392,13 @@ buildSuperCluster(CalibClusterPtr& seed,
 	<< std::endl;
     }    
   }
+
+  //temporary fix for HGCal issue
+  if (not_clustered == clusters.begin()) {
```
Comment (slava77) [inline RecoEcal/EgammaClusterAlgos/src/PFECALSuperClusterAlgo.cc]: IIUC, the problem affects cases of this code running on HGCAL inputs.
is it obvious to give no effect on regular ECAL clustering?
If not, please add a check that this is HGCAL.

Comment (kpedro88) [reply RecoEcal/EgammaClusterAlgos/src/PFECALSuperClusterAlgo.cc]: If this ever happened in ECAL, we would have noticed the algorithm crashing before now.

My understanding from @rovere's investigation is that this can only happen for HGCal because of the different meaning of calibrated vs uncalibrated energy.

**Comment by kpedro88** [comment]
+1

**Comment by rovere** [comment]
As I suspected, there was something else fishy going on. Maybe this PR and the protection could stay as they are since they won't harm for sure.
Expect another PR to fix the real bug related to the realistic SimClusters within the day.

@malgeri

File: RecoEcal/EgammaClusterAlgos/src/PFECALSuperClusterAlgo.cc (hunk 1 above)
Comment (slava77) [reply RecoEcal/EgammaClusterAlgos/src/PFECALSuperClusterAlgo.cc]: @lgray  [as the main developer of this in the past]
please check if this change is reasonable

**Comment by kpedro88** [comment]
@rovere time is of the essence...

**Comment by rovere** [comment]
@kpedro88 we know it and are working on the fix + new calibration within the day.

**Comment by rovere** [comment]
As said, this could be considered as a protection/bug fix, if you like.
The other PR will fix the problem that surfaced this bug. In this regard, this PR, from my side, can go in as is right now.

**Comment by kpedro88** [comment]
unhold

File: RecoEcal/EgammaClusterAlgos/src/PFECALSuperClusterAlgo.cc (hunk 1 above)
Comment (slava77) [reply RecoEcal/EgammaClusterAlgos/src/PFECALSuperClusterAlgo.cc]: So, if I understand correctly, the buildSuperCluster is supposed to be called only on seedable clusters.
The proposed change is a quiet escape from a bug introduced by the caller where the selection of seedable clusters contradicts any ability to match the list with a seed.
This is not good.

The solution appears to be OK for the current use cases of buildSuperCluster, but it opens up the property of this method to handle unseedable list of clusters, which should not be exercised.
I propose to add a config parameter (e.g. ```dropUnseedable ``` set to false by default ) and set it to true for HGCAL and also add a LogError or even an exception if dropUnseedable is false and we end up here.

Comment (rovere) [reply RecoEcal/EgammaClusterAlgos/src/PFECALSuperClusterAlgo.cc]: Ciao @slava77 the real fix in the realistic simcluster energy assignment will automatically fix this issue and will not crash. Having said that, the code as is is still fragile since, as I explained in the text reported above by @kpedro88 the conditions to be seedable are not safe for the mathematical operations that will follow in doing the partitioning of the clusters.
So, if you prefer, the code is not protected but feeding it with decent cluster won't crash.
Whether we want to consider it as a bug or a feature is, in a sense, somewhat personal.

**Comment by kpedro88** [comment]
@slava77 exception output on the failing event mentioned in the PR description when `dropUnseedable` is set to false for HGCal:
```
Begin processing the 1st record. Run 1, Event 2959, LumiSection 30 at 12-Sep-2017 11:53:29.422 CDT
----- Begin Fatal Exception 12-Sep-2017 11:53:36 CDT-----------------------
An exception of category 'PFECALSuperClusterAlgo::buildSuperCluster' occurred while
   [0] Processing  Event run: 1 lumi: 30 event: 2959 stream: 0
   [1] Running path 'prevalidation_step1'
   [2] Prefetching for module MultiTrackValidator/'trackValidator'
   [3] Prefetching for module JetTracksAssociationToTrackRefs/'cutsRecoTracksAK4PFJets'
   [4] Prefetching for module JetTracksAssociatorExplicit/'ak4JetTracksAssociatorExplicitAll'
   [5] Prefetching for module FastjetJetProducer/'ak4PFJets'
   [6] Prefetching for module PFLinker/'particleFlow'
   [7] Prefetching for module PFCandidateListMerger/'particleFlowTmp'
   [8] Prefetching for module PFProducer/'particleFlowTmpBarrel'
   [9] Prefetching for module PFBlockProducer/'particleFlowBlock'
   [10] Prefetching for module PFElecTkProducer/'pfTrackElec'
   [11] Prefetching for module GsfTrackProducer/'electronGsfTracks'
   [12] Prefetching for module CkfTrackCandidateMaker/'electronCkfTrackCandidates'
   [13] Prefetching for module ElectronSeedMerger/'electronMergedSeeds'
   [14] Prefetching for module ElectronSeedProducer/'ecalDrivenElectronSeeds'
   [15] Calling method for module PFECALSuperClusterProducer/'particleFlowSuperClusterHGCal'
Exception Message:
Cluster is not seedable!
    Non-Clustered cluster: e = 0.0662978 eta = -1.71095 phi = -1.38067
----- End Fatal Exception -------------------------------------------------
```

**Comment by slava77** [comment]
@kpedro88 
thanks for the update.
The behavior looks reasonable with the "false".

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by kpedro88** [comment]
+1

File: RecoEcal/EgammaClusterAlgos/src/PFECALSuperClusterAlgo.cc [hunk 2]
```diff
@@ -392,6 +392,22 @@ buildSuperCluster(CalibClusterPtr& seed,
 	<< std::endl;
     }    
   }
+
+  //temporary fix for HGCal issue
```
Comment (slava77) [inline RecoEcal/EgammaClusterAlgos/src/PFECALSuperClusterAlgo.cc]: the comment is now out of sync. Based on the discussion in this PR, the fix is no longer temporary.

**Comment by slava77** [comment]
@cmsbuild please test
to have it all green in the end

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/20458	57e3745
- changes to make PFECALSuperClusterAlgo::buildSuperCluster more strict by default and add a configuration switch ```dropUnseedable``` to allow somewhat inconsistent situations to cover corner cases where the current logic of selection of a seedable cluster and a match to a seed is sometimes inconsistent.
- jenkins tests pass and comparisons show no differences.

The claim is that the unseedable cluster situation never happens in non-HGCAL clustering, otherwise we'd see crashes in production in ECAL clustering [but we have not seen any related to this part]. So, it sounds safe to be included and backported to 93X.

**Comment by kpedro88** [comment]
+1

**Comment by davidlange6** [comment]
+1
